### PR TITLE
Restructure different file system implementations to separate files

### PR DESCRIPTION
**Pull request by GuanLuo**

### CODE REVIEW

**Comment by kthui** [comment]
I wonder if it might be clearer to add one more layer into the filesystem directory, separating the Triton's API interface and individual file system interface. For example:
```
src
|
`--- filesystem
     |
     `--- api.h
     |
     `--- api.cc
     |
     `--- implementations
          |
          `--- common.h
          |
          `--- as.h
          |
          `--- gcs.h
          |
          `--- local.h
          |
          `--- s3.h
```
If Triton wants to call any filesystem API, everything will be inside the `api.h`. If any new filesystem implementation is to be added, they will be inside the `implementations` directory.

File: src/filesystem/as.h [hunk 1]
```diff
@@ -0,0 +1,470 @@
...
+}
+
+Status
+ASFileSystem::DeletePath(const std::string& path)
+{
+  return Status(
+      Status::Code::UNSUPPORTED, "Delete path operation not yet implemented");
+}
+
+}}  // namespace triton::core
```
Comment (kthui) [inline src/filesystem/as.h:470]: Maybe add an empty line at the end of file?

File: src/filesystem/gcs.h [hunk 2]
```diff
@@ -0,0 +1,497 @@
...
+}
+
+Status
+GCSFileSystem::DeletePath(const std::string& path)
+{
+  return Status(
+      Status::Code::UNSUPPORTED, "Delete path operation not yet implemented");
+}
+
+}}  // namespace triton::core
```
Comment (kthui) [inline src/filesystem/gcs.h:497]: Maybe add an empty line at the end of file?

File: src/filesystem/local.h [hunk 3]
```diff
@@ -0,0 +1,348 @@
...
+      RETURN_IF_ERROR(DeletePath(JoinPath({path, content})));
+    }
+    rmdir(path.c_str());
+  } else {
+    remove(path.c_str());
+  }
+  return Status::Success;
+}
+
+}}  // namespace triton::core
```
Comment (kthui) [inline src/filesystem/local.h:348]: Maybe add an empty line at the end of file?

File: src/filesystem/s3.h [hunk 4]
```diff
@@ -0,0 +1,725 @@
...
+}
+
+Status
+S3FileSystem::DeletePath(const std::string& path)
+{
+  return Status(
+      Status::Code::UNSUPPORTED, "Delete path operation not yet implemented");
+}
+
+}}  // namespace triton::core
```
Comment (kthui) [inline src/filesystem/s3.h:725]: Maybe add an empty line at the end of file?

**Review by Tabrizian — APPROVED** [review]

File: src/CMakeLists.txt [hunk 5]
```diff
@@ -116,7 +116,7 @@ set(
   dynamic_batch_scheduler.cc
   ensemble_scheduler.cc
   ensemble_utils.cc
-  filesystem.cc
```
[the remark this answers is not in the document]
Comment (GuanLuo) [reply src/CMakeLists.txt:119]: They are included in `api.cc`, so in compiler's view `api.cc` is the same as the giant `filesystem.cc` file previously, but broken in several files in development.

**Review by kthui — APPROVED** [review]

**Pull request merged**